elytron/WFCORE-4058-automatic-keystore-detection.adoc

```diff
@@ -0,0 +1,92 @@
+= Automatic detection of file based KeyStore types.
+:author:            Maritn Mazanek
+:email:             mmazanek@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+* If a user does not specify the type of a file based KeyStore can we auto detect the type?
+
+* There is most likely a finite set of types we would want to automatically recognise, possibly we could inspect the
+contents ourselves to detect different types - alternative attempt initialisation of different KeyStore objects until one succeeds. With the latter we would need to be careful one doesn't silently succeed when it should have failed.
```

Comment:
Personally I consider trying loading of keystore more robust approach for this. To be independent of implementation details of keystore types. Also there can be optimalisation try to load only types which are discoverable by JCA. It means try only keystores which we have a chance to load.

Also this detection should be performed ideally only once.

Reply:
The load at the end will become the final accurate check - deliberately attempting to initialise a KeyStore multiple times seems quite inefficient - each implementation may have it's own initialisation to perform before the file is read which may be preparation to decrypt an encrypted stream, we then need to read the file in once for each KeyStore we try.  We then need to handle the failure conditions leaving the failed attempt to be garbage collected.

If we reduced the scope to just JKS, JCEKS, and PKCS12 they seem to be coming out with the more predicable signature.

Reply:
Yeah, it would probably not be so good to try to discover encrypted keystores, so leave the distinguishable and select the type based on that? If that wouldn't succeed we can simply throw exception and there would be no need for another discovery code.